### ISSUE TITLE
IsValidLocationForInlining no longer considered when running await continuations

### ISSUE THREAD

**Issue by caozhiyuan** [dotnet/corefx#27675]
when up to 2.1 preview1  run app in cmd dotnet run -c release -f netcoreapp2.1
TaskCompletionSource TrySetResult  be blocked
netcoreapp2.0 is ok

```C#
 public sealed class TaskCompletionSource : TaskCompletionSource<int>
    {
        public static readonly TaskCompletionSource Void = CreateVoidTcs();

        public TaskCompletionSource(object state)
            : base(state)
        {
        }

        public TaskCompletionSource()
        {
        }

        public bool TryComplete()
        {
           return this.TrySetResult(0);
        }

        public void Complete() => this.SetResult(0);

        // todo: support cancellation token where used
        public bool SetUncancellable() => true;

        public override string ToString() => "TaskCompletionSource[status: " + this.Task.Status.ToString() + "]";

        static TaskCompletionSource CreateVoidTcs()
        {
            var tcs = new TaskCompletionSource();
            tcs.TryComplete();
            return tcs;
        }
    }
```

**Comment by stephentoub** [comment]
What's the issue? I'm not understanding.

**Comment by caozhiyuan** [comment]
@stephentoub  it's weird, 
```  C#
         Console.WriteLine("111");
            try
            {

                return this.TrySetResult(0);
            }
            finally
            {
                Console.WriteLine("22222222");
            }
``` 
i try to write a sample repo, in netcoreapp2.0 can print 22222222  but 2.1 run in cmd TrySetResult blocked sometimes.  [https://github.com/Azure/DotNetty/blob/ec9820d74716c8a6d84954ad1f46a7eec9f20eef/src/DotNetty.Transport/Bootstrapping/AbstractBootstrap.cs#L296](url)

**Comment by caozhiyuan** [comment]
it may be not corefx issue

**Comment by caozhiyuan** [comment]
![qq 20180303145734](https://user-images.githubusercontent.com/3415285/36931561-654c0e1c-1ef3-11e8-922a-c7abcdbfde30.png)
@stephentoub  when it be blocked?

**Comment by caozhiyuan** [comment]
@stephentoub  https://github.com/Azure/DotNetty/blob/dev/src/DotNetty.Transport/Bootstrapping/AbstractBootstrap.cs#L232     IChannel channel = await this.InitAndRegisterAsync()  after this method 2.1 thread be loop thread (because loop thread call TaskCompletionSource TrySetResult )  but net coreapp 2.0 is work thread

**Comment by caozhiyuan** [comment]
this is sample @stephentoub 
``` C#
    class Program
    {
        static ExecutorTaskScheduler scheduler = new ExecutorTaskScheduler();

        static void Main(string[] args)
        {
            Console.WriteLine("start");
            Thread thread = new Thread(Start)
            {
                IsBackground = true
            };
            thread.Start();

            Test2().Wait();

            Console.WriteLine("done");
            Console.ReadLine();
        }

        public static async Task Test2()
        {
            await SendRequest();
        }

        public static Task SendRequest()
        {
            var promise = new TaskCompletionSource();

            Queue.Enqueue(new MyRunnable((p) =>
            {
                Thread.Sleep(10);
                ((TaskCompletionSource)p).TryComplete();
            }, promise));
            autoResetEvent.Set();

            return promise.Task;          
        }

        private static AutoResetEvent autoResetEvent = new AutoResetEvent(true);

        private static void Start()
        {
            Task.Factory.StartNew(
                () =>
                {
                    try
                    {
                        while (true)
                        {
                            if (Queue.TryDequeue(out var result))
                            {
                                try
                                {
                                    result.Run();
                                }
                                catch (Exception e)
                                {
                                    Console.WriteLine(e);
                                }
                            }
                            else
                            {
                                autoResetEvent.WaitOne();
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine(ex);
                    }
                },
                CancellationToken.None,
                TaskCreationOptions.None,
                scheduler);
        }

        private static ConcurrentQueue<IRunnable> Queue = new ConcurrentQueue<IRunnable>();

    }

    public class MyRunnable : IRunnable
    {
        private Action<object> _action;
        private object _state;

        public MyRunnable(Action<object> action, object state)
        {
            _action = action;
            _state = state;
        }

        public void Run()
        {
            _action(_state);
        }
    }

    public interface IRunnable
    {
        void Run();
    }

    public sealed class ExecutorTaskScheduler : TaskScheduler
    {
        protected override void QueueTask(Task task)
        {
            this.TryExecuteTask(task);
        }

        protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
        {
            return this.TryExecuteTask(task);
        }

        protected override IEnumerable<Task> GetScheduledTasks() => null;

        protected override bool TryDequeue(Task task) => false;
    }

    public sealed class TaskCompletionSource : TaskCompletionSource<int>
    {
        public static readonly TaskCompletionSource Void = CreateVoidTcs();

        public TaskCompletionSource(object state)
            : base(state)
        {
        }

        public TaskCompletionSource()
        {
        }

        public bool TryComplete()
        {
            return this.TrySetResult(0);
        }

        public void Complete() => this.SetResult(0);
        

        public bool SetUncancellable() => true;

        public override string ToString() => "TaskCompletionSource[status: " + this.Task.Status.ToString() + "]";

        static TaskCompletionSource CreateVoidTcs()
        {
            var tcs = new TaskCompletionSource();
            tcs.TryComplete();
            return tcs;
        }
    }
```

**Comment by stephentoub** [comment]
There are various oddities in that sample, but it runs fine for me on 2.1.  When you see TrySetResult blocks, have you attached a debugger to see what's on that call stack when that happens?  TrySetResult will invoke any synchronous continuations that have registered with the Task associated with that TaskCompletionSource, so it's possible you've created some kind of deadlock where you're invoking continuations that are then blocked on the thread calling TrySetResult doing something else, and such things could be subject to timing differences, which could explain why you might see things in debug vs release or in 2.0 vs 2.1.  One of the oddities I mentioned is that your scheduler's QueueTask is actually synchronously executing the provided Task as well, and schedulers are inherited, so any code that runs in one of the queued operations will inherit that behavior when calling Run, and thus any tasks created as part of that will end up getting invoked synchronously when scheduled instead of asynchronously; that could also lead to deadlocks.

**Comment by caozhiyuan** [comment]
add debug breakpoint after  await ;  view thread stack , 2.1 diff with 2.0 . 
.net core 2.0 ![qq 20180303184800](https://user-images.githubusercontent.com/3415285/36935201-acb85968-1f2f-11e8-8da7-a23b9b76255f.png)

.net core 2.1 ![qq 20180303185122](https://user-images.githubusercontent.com/3415285/36935204-b2128712-1f2f-11e8-918b-e572eb360560.png)

**Comment by stephentoub** [comment]
Right, so in the 2.1 view, I see on the call stack what looks like the call to TryComplete? And then there's a bunch of stuff that it's calling in other libraries due to the continuations. So something there is causing it to hang. I don't believe this is an issue with TaskCompletionSource.

Does the standalone repro you shared hang for you on 2.1? These images are obviously not from that repro.

Thanks.

**Comment by caozhiyuan** [comment]
@stephentoub  change code like this  it work same as netcoreapp2.0 (
i still  can't get a standalone repro to explain it.)

```C#

    public sealed class TaskCompletionSource : TaskCompletionSource<int>
    {
        public static readonly TaskCompletionSource Void = CreateVoidTcs();

        public TaskCompletionSource(object state)
            : base(state, TaskCreationOptions.RunContinuationsAsynchronously)
        {
        }

        public TaskCompletionSource() : base(TaskCreationOptions.RunContinuationsAsynchronously)
        {
        }

        public bool TryComplete()
        {
            return this.TrySetResult(0);
        }

        public void Complete() => this.SetResult(0);
        

        public bool SetUncancellable() => true;

        public override string ToString() => "TaskCompletionSource[status: " + this.Task.Status.ToString() + "]";

        static TaskCompletionSource CreateVoidTcs()
        {
            var tcs = new TaskCompletionSource();
            tcs.TryComplete();
            return tcs;
        }
    }
```

**Comment by yyjdelete** [comment]
Maybe an repro. Run it with `netcoreapp2.0`, `2.1` with `Task.s_asyncDebuggingEnabled=false`(default) and `2.1` with `Task.s_asyncDebuggingEnabled=true`(default in VS);
You can see different StackTrace with 2.0 and 2.1(with false), and in 2.1 `Do something else.` will never show before press Enter.
```cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    class Program
    {
        private static ExecutorTaskScheduler scheduler;

        static void Main(string[] args)
        {
            //typeof(Task).GetField("s_asyncDebuggingEnabled", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic)?.SetValue(null, true);

            scheduler = new ExecutorTaskScheduler();
            MainAsync().Wait();
        }

        private static async Task MainAsync()
        {
            Console.WriteLine("MainAsync-Pre->" + Thread.CurrentThread.ManagedThreadId);
            await X1();
            Console.WriteLine("MainAsync-Post, and block->" + Thread.CurrentThread.ManagedThreadId);
            //Console.WriteLine(Environment.StackTrace);
            Console.ReadLine();

        }

        private static Task X1()
        {
            Console.WriteLine("X1->" + Thread.CurrentThread.ManagedThreadId);
            var tmp = new TaskCompletionSource<int>();
            Task.Factory.StartNew(() =>
            {
                Console.WriteLine("Task->" + Thread.CurrentThread.ManagedThreadId);
                tmp.TrySetResult(0);
                Console.WriteLine("Do something else.");
                while (true)
                {
                    Thread.Sleep(100);
                }
            }, CancellationToken.None, TaskCreationOptions.LongRunning, scheduler);
            return tmp.Task;
        }

        public sealed class ExecutorTaskScheduler : TaskScheduler
        {

            protected override void QueueTask(Task task)
            {
                ThreadPool.QueueUserWorkItem((state) =>
                {
                    this.TryExecuteTask(task);
                });
            }

            protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
            {
                return false;
            }

            protected override IEnumerable<Task> GetScheduledTasks() => null;

            protected override bool TryDequeue(Task task) => false;
        }
    }
}
```

**Comment by yyjdelete** [comment]
@stephentoub Maybe related to #14178 according the StackTrace?

**Comment by yyjdelete** [comment]
Update the test code to see the StackTrace like the original case.
<details>
<summary>Test code</summary>

```cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Reflection;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    class Program
    {
        private static ExecutorTaskScheduler scheduler;

        static void Main(string[] args)
        {
            typeof(Task).GetTypeInfo().GetField("s_asyncDebuggingEnabled", BindingFlags.Static | BindingFlags.NonPublic)?.SetValue(null, args.Length != 0);

            scheduler = new ExecutorTaskScheduler();
            scheduler.Start();

            MainAsync2().Wait();
            Thread.Sleep(100);
        }

        private static async Task MainAsync2()
        {
            await MainAsync();
            Console.WriteLine(Environment.StackTrace);
            Console.ReadLine();
            Console.WriteLine("<press enter>");
        }

        private static async Task MainAsync()
        {
            Console.WriteLine("MainAsync-Pre->" + Thread.CurrentThread.ManagedThreadId);
            await X1();
            Console.WriteLine("MainAsync-Post, and block->" + Thread.CurrentThread.ManagedThreadId);
        }

        private static Task X1()
        {
            Console.WriteLine("X1->" + Thread.CurrentThread.ManagedThreadId);
            var tmp = new TaskCompletionSource<int>();
            Task.Factory.StartNew(() =>
            {
                Console.WriteLine("Task1->" + Thread.CurrentThread.ManagedThreadId);
                tmp.TrySetResult(0);
            }, CancellationToken.None, TaskCreationOptions.None, scheduler);
            new Thread(() =>
            {
                Thread.Sleep(100);
                Task.Factory.StartNew(() =>
                {
                    Console.WriteLine("Loop is still alive->" + Thread.CurrentThread.ManagedThreadId);
                }, CancellationToken.None, TaskCreationOptions.None, scheduler);
            }).Start();
            return tmp.Task;
        }

        public sealed class ExecutorTaskScheduler : TaskScheduler
        {
            private ConcurrentQueue<Task> queue;
            private int loopThread = -1;

            private bool IsInLoop => Thread.CurrentThread.ManagedThreadId == this.loopThread;

            private void Loop()
            {
                this.queue = new ConcurrentQueue<Task>();
                this.loopThread = Thread.CurrentThread.ManagedThreadId;
                Console.WriteLine("Start Loop->" + Thread.CurrentThread.ManagedThreadId);
                while (true)
                {
                    while (queue.TryDequeue(out var task))
                    {
                        this.TryExecuteTask(task);
                    }
                    Thread.Sleep(10);
                }
            }

            public void Start()
            {
                new Thread(Loop) { IsBackground = true }.Start();
            }

            protected override void QueueTask(Task task)
            {
                queue.Enqueue(task);
            }

            protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued)
            {
                if (taskWasPreviouslyQueued || !IsInLoop)
                {
                    return false;
                }

                return this.TryExecuteTask(task);
            }

            protected override IEnumerable<Task> GetScheduledTasks() => null;

            protected override bool TryDequeue(Task task) => false;
        }
    }
}
```

</details>

---

`dotnet run -f netcoreapp2.0`
<details>

```
Start Loop->3
MainAsync-Pre->1
X1->1
Task1->3
MainAsync-Post, and block->5
   at System.Environment.get_StackTrace()
   at ConsoleApp1.Program.<MainAsync2>d__2.MoveNext() in Z:\tst1\Program.cs:line 28
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetExistingTaskResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(Task`1 completedTask)
   at ConsoleApp1.Program.<MainAsync>d__3.MoveNext() in Z:\tst1\Program.cs:line 38
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
Loop is still alive->3

<press enter>
```
</details>

---

`dotnet run -f netcoreapp2.1`(with `Task.s_asyncDebuggingEnabled=false`)
See `Loop is still alive` is blocked by `ReadLine`
<details>

```
Start Loop->3
MainAsync-Pre->1
X1->1
Task1->3
MainAsync-Post, and block->3
   at System.Environment.get_StackTrace()
   at ConsoleApp1.Program.MainAsync2() in Z:\tst1\Program.cs:line 28
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetExistingTaskResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at ConsoleApp1.Program.MainAsync() in Z:\tst1\Program.cs:line 38
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Threading.Tasks.TaskCompletionSource`1.TrySetResult(TResult result)
   at ConsoleApp1.Program.<>c__DisplayClass4_0.<X1>b__0() in Z:\tst1\Program.cs:line 47
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry()
   at ConsoleApp1.Program.ExecutorTaskScheduler.Loop() in Z:\tst1\Program.cs:line 76
   at System.Threading.Thread.ThreadMain_ThreadStart()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)

<press enter>
Loop is still alive->3
```
</details>

---

`dotnet run -f netcoreapp2.1 arg0`(with `Task.s_asyncDebuggingEnabled=true`)
<details>

```
Start Loop->3
MainAsync-Pre->1
X1->1
Task1->3
MainAsync-Post, and block->5
   at System.Environment.get_StackTrace()
   at ConsoleApp1.Program.MainAsync2() in Z:\tst1\Program.cs:line 28
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.<>c.<OutputWaitEtwEvents>b__12_0(Action innerContinuation, Task innerTask)
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetExistingTaskResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at ConsoleApp1.Program.MainAsync() in Z:\tst1\Program.cs:line 38
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.<>c.<OutputWaitEtwEvents>b__12_0(Action innerContinuation, Task innerTask)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
Loop is still alive->3

<press enter>
```
</details>

**Comment by caozhiyuan** [comment]
@stephentoub  take a look above

**Comment by stephentoub** [comment]
> You can see different StackTrace with 2.0 and 2.1(with false), and in 2.1(with false) Do something else. will never show before press Enter.

I do not see any meaningful differences here between 2.0 and 2.1.  What is it you believe this code shows?  There are race conditions in the sample code, so different runs will produce different stack traces, regardless of whether 2.0 or 2.1 is being used, and regardless of whether that s_asyncDebuggingEnabled flag is set.

> It used to call AwaitTaskContinuation.RunOrScheduleAction() with check for IsValidLocationForInlining, but not now.

As far as I see, this has nothing to do with 2.0 vs 2.1, but rather it has to do with a race condition in your code.  You're creating a Task.Factory.StartNew task scheduled to "scheduler", and that task is calling TrySetResult on a TaskCompletionSource task that's returned and awaited.  If the queued Task.Factory.StartNew task runs and completes prior to the await happening, then the await will happen on an already completed task.  But if the Task.Factory.StartNew task runs and completes after the await has already been initiated, then when you call TrySetResult, it'll invoke the continuation, which will check to see whether the current task scheduler matches the one that was present when the await happened; it won't match, because the default scheduler was there for the await and the TrySetResult is being called with your custom "scheduler", and so it'll force the continuation to be queued asynchronously.  You can see how things are affected by, for example, passing TaskCreationOptions.HideScheduler as the options to your Task.Factory.StartNew task.

As I noted, I still do not see any bug here in TaskCompletionSource.

**Comment by yyjdelete** [comment]
@stephentoub Already add an `Thread.Sleep(100)` before `TrySetResult` in local code, and directly call `MainAsync` instead of `MainAsync2` for simple debug.

The call of await in netcoreapp2.1 with false(in Thread 1, no scheduler)
![await](https://user-images.githubusercontent.com/1819074/36976852-fa974ab4-20b9-11e8-840f-ecb89c0b4553.png)
The call of continue in netcoreapp2.1 with false(in Thread 3, with scheduler but never check)
![continue](https://user-images.githubusercontent.com/1819074/36976856-fd553e28-20b9-11e8-9662-9ca8cc79cd56.png)

With netcoreapp2.0 or 2.1 with true(almost the same as 2.0), it will run into an different case with check for `IsValidLocationForInlining`
![image](https://user-images.githubusercontent.com/1819074/36977076-e880951e-20ba-11e8-8360-9c4b45304d2e.png)
![image](https://user-images.githubusercontent.com/1819074/36977084-efc81068-20ba-11e8-8522-584cc514bbf5.png)
![image](https://user-images.githubusercontent.com/1819074/36977104-fe06031a-20ba-11e8-838a-866e3796cb2d.png)

**Comment by stephentoub** [comment]
I'm sorry, but I don't understand what it is you're trying to do show me. What is the exact issue you're highlighting? What do you believe is a bug here?

**Comment by yyjdelete** [comment]
I'm not sure whether it's an bug or expected behavior, but the behavior does changes in 2.1preview. And according the doc of `IsValidLocationForInlining` continuations shouldn't be inline if the current thread has a non-null SynchronizationContext or a non-default scheduler. (but maybe an internal behavior since no public docment for it?)

This will makes some slow usercode leak to loopThread(2.1) instead of run in threadpool(2.0) when await and call TrySetResult in [DotNetty](https://github.com/Azure/DotNetty). Not sure whether this also affact some program with UI Thread.

**Comment by caozhiyuan** [comment]
Task.s_asyncDebuggingEnabled true go task.SetContinuationForAwait otherwise task.UnsafeSetContinuationForAwait 
this 2 methods have diff logic 
```C#
        /// <summary>Schedules the continuation onto the <see cref="System.Threading.Tasks.Task"/> associated with this <see cref="TaskAwaiter"/>.</summary>
        /// <param name="task">The task being awaited.</param>
        /// <param name="continuation">The action to invoke when the await operation completes.</param>
        /// <param name="continueOnCapturedContext">Whether to capture and marshal back to the current context.</param>
        /// <param name="flowExecutionContext">Whether to flow ExecutionContext across the await.</param>
        /// <exception cref="System.ArgumentNullException">The <paramref name="continuation"/> argument is null (Nothing in Visual Basic).</exception>
        /// <exception cref="System.NullReferenceException">The awaiter was not properly initialized.</exception>
        /// <remarks>This method is intended for compiler user rather than use directly in code.</remarks>
        internal static void UnsafeOnCompletedInternal(Task task, IAsyncStateMachineBox stateMachineBox, bool continueOnCapturedContext)
        {
            Debug.Assert(stateMachineBox != null);

            // If TaskWait* ETW events are enabled, trace a beginning event for this await
            // and set up an ending event to be traced when the asynchronous await completes.
            if (TplEtwProvider.Log.IsEnabled() || Task.s_asyncDebuggingEnabled)
            {
                task.SetContinuationForAwait(OutputWaitEtwEvents(task, stateMachineBox.MoveNextAction), continueOnCapturedContext, flowExecutionContext: false);
            }
            else
            {
                task.UnsafeSetContinuationForAwait(stateMachineBox, continueOnCapturedContext);
            }
        }
```

**Comment by caozhiyuan** [comment]
this commit add it
SHA-1: 9a0630152c4339324ba9ac0fa886f9fb4916e759 Avoid async method delegate allocation.
@stephentoub  should this 2 method behavior same ?

**Comment by yyjdelete** [comment]
@stephentoub At least this is a breaking change that with different behavior compare to netfx/netcoreapp2.0, and can't be debuged in VS. And should be included in the changelog, and update the document of `IsValidLocationForInlining` to match the new behavior.

**Comment by stephentoub** [comment]
> ...but s_asyncDebuggingEnabled cause different behavior , i think it's bad.
> ...the behavior does changes in 2.1preview, and it's an breaking change

How?  What is the behavioral change?  Can you please be very specific and point to what is the difference?  I've yet to see a repro that I can run that highlights the difference; all of the samples that have been provided thus far exhibit the same behavior for me between 2.0 and 2.1, the only difference being race conditions that manifest on both and yield multiple behaviors but with multiple behaviors exhibited on both.

If there's a bug I would like to fix it, but I still do not understand what the issue is you're trying to highlight.

You're saying there's a functional difference in behavior between the internal SetContinuationForAwait and the internal UnsafeSetContinuationForAwait?  What is that difference?  The contain different code but they should exhibit the exact same visible behavior.

Do you see a difference if you use the code that's in master?  What 2.1 version are you using?

> and update the document of IsValidLocationForInlining to match the new behavior.

IsValidLocationForInlining is an internal method and it has not changed one bit between 2.0 and 2.1:
2.0: https://github.com/dotnet/coreclr/blob/release/2.0.0/src/mscorlib/src/System/Threading/Tasks/TaskContinuation.cs#L612-L626
2.1: https://github.com/dotnet/coreclr/blob/release/2.1/src/mscorlib/src/System/Threading/Tasks/TaskContinuation.cs#L612-L626

**Comment by caozhiyuan** [comment]
@stephentoub 
2.1 s_asyncDebuggingEnabled true or 2.0 , go https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L3272

2.1 s_asyncDebuggingEnabled  false , go 
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L3283

diff is use ThreadPool or not , it looks driect invoke will cause dotnetty deadlock. 
and if we add RunContinuationsAsynchronously also can slove deadlock problem.

**Comment by stephentoub** [comment]
> 2.1 s_asyncDebuggingEnabled true or 2.0 , go https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L3272
> 2.1 s_asyncDebuggingEnabled false , go
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L3283

Ok, thank you, I now understand what you're getting at and why you were mentioning IsValidLocationForInlining.  It's not that IsValidLocationForInlining has changed; it hasn't.  What's changed is that we're now bypassing the IsValidLocationForInlining check, and that check is important for the scenario.  That's something I can look at.
